stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                        n,
+                        &next_wanted_tenures[n]
+                    );
+                    break;
+                }
+            }
+            let wt_start = if i < wanted_tenures.len() {
+                let Some(wt) = wanted_tenures.get(i) else {
+                    break;
+                };
+                wt
+            } else {
+                let Some(wt) = next_wanted_tenures.get(n) else {
+                    break;
+                };
+                wt
+            };
+            test_debug!("next start tenure is {:?}", &wt_start);
+
+            // find the next 1-bit after that -- corresponds to the tenure-end block ID.
+            // `k` necessarily points the tenure in `next_wanted_tenures` which corresponds to the
+            // tenure after the previously-found tenure (either `wanted_tenures[i]` or
+            // `next_wanted_tenures[n]`, depending on the blockchain structure).
+            let mut k = if next {
+                // start block is in `next_wanted_tenures` (at `n`), so search for the wanted
+                // tenure whose bit is after `n`
+                n + 1
+            } else {
+                // start block is in `wanted_tenures`, and it's the last tenure that has a 1-bit in
+                // `wanted_tenures`. Start searching `next_wanted_tenures`.
+                0
+            };
+
+            while k < next_wanted_tenures.len() {
+                let bit = u16::try_from(k).expect("FATAL: more sortitions than u16::MAX");
+                if !next_invbits.get(bit).unwrap_or(false) {
+                    k += 1;
+                    continue;
+                }
+
+                // k now points to the item in wanted_tenures with the tenure-send block ID for
+                // `ch`.
+                test_debug!("next end tenure is k={} {:?}", k, &next_wanted_tenures[k]);
+                break;
+            }
+            let Some(wt_end) = next_wanted_tenures.get(k) else {
+                break;
+            };
+
+            let mut tenure_start_end = TenureStartEnd::new(
+                wt.tenure_id_consensus_hash.clone(),
+                wt_start.winning_block_id.clone(),
+                wt_end.winning_block_id.clone(),
+                rc,
+                pox_constants
+                    .block_height_to_reward_cycle(first_burn_height, wt_start.burn_height)
+                    .expect("FATAL: tenure from before system start"),
+                wt.processed,
+            );
+            tenure_start_end.fetch_end_block = true;
+            test_debug!(
+                "i={}, k={}, n={}, len={}, next_len={}; {:?}",
+                i,
+                k,
+                n,
+                wanted_tenures.len(),
+                next_wanted_tenures.len(),
+                &tenure_start_end
+            );
+            tenure_block_ids.insert(wt.tenure_id_consensus_hash.clone(), tenure_start_end);
+        }
+        Some(tenure_block_ids)
+    }
+}
```

Comment:
This method is really obtuse. `i` appears to be the iterating variable, but its modified in both the outer loop, and an inner loop, before being restored to the value it was at while iterating in the outer loop. And then there's a whole additional outer loop with the same structuring. I think this is a classic case where `find()` would be really really clarifying:

```suggestion
    pub fn from_inventory(
        rc: u64,
        wanted_tenures: &[WantedTenure],
        next_wanted_tenures: Option<&[WantedTenure]>,
        pox_constants: &PoxConstants,
        first_burn_height: u64,
        invs: &NakamotoTenureInv,
    ) -> Option<AvailableTenures> {
        // if bit i is set, that means that the tenure data for the ith tenure in the sortition
        // history was present.  But given that block-commits commit to the start block of the
        // parent tenure, the start-block ID for tenure i would be the StacksBlockId for the
        // next-available tenure.  Its end-block ID would be the StacksBlockId for the
        // next-available tenure after that.
        let invbits = invs.tenures_inv.get(&rc)?;
        let mut tenure_block_ids = AvailableTenures::new();

        let mut last_tenure = 0;
        let mut last_tenure_ch = None;
        for (i, wt) in wanted_tenures.iter().enumerate() {
            // advance to next tenure-start sortition
            let bit = u16::try_from(i).expect("FATAL: more sortitions than u16::MAX");
            if !invbits.get(bit).unwrap_or(false) {
                test_debug!("i={} bit not set", i);
                continue;
            }

            // the last tenure we'll consider
            last_tenure = i;

            // find next 1-bit -- corresponds to tenure-start block ID
            let Some(wt_start_ix) = (i..wanted_tenures.len()).find(|j| {
                let bit_index = u16::try_from(*j).expect("FATAL: more sortitions than u16::MAX");
                invbits.get(bit_index).unwrap_or(false)
            }) else {
                test_debug!("Out of wanted_tenures");
                break;
            };
            let Some(wt_start) = wanted_tenures.get(wt_start_ix) else {
                test_debug!("i={} no start wanted tenure", i);
                break;
            };

            // find the next 1-bit after that -- corresponds to the tenure-end block ID.
            let Some(wt_end_ix) = ((wt_start_ix + 1)..wanted_tenures.len()).find(|j| {
                let bit_index = u16::try_from(*j).expect("FATAL: more sortitions than u16::MAX");
                invbits.get(bit_index).unwrap_or(false)
            }) else {
                test_debug!("Out of wanted_tenures");
                break;
            };
            let Some(wt_end) = wanted_tenures.get(wt_end_ix) else {
                test_debug!("i={} no start wanted tenure", i);
                break;
            };

            let tenure_start_end = TenureStartEnd::new(
                wt.tenure_id_consensus_hash.clone(),
                wt_start.winning_block_id.clone(),
                wt_end.winning_block_id.clone(),
                rc,
                rc,
                wt.processed,
            );

            last_tenure_ch = Some(wt.tenure_id_consensus_hash.clone());
            tenure_block_ids.insert(wt.tenure_id_consensus_hash.clone(), tenure_start_end);
        }

        let Some(next_wanted_tenures) = next_wanted_tenures else {
            // nothing more to do
            test_debug!("No next_wanted_tenures");
            return Some(tenure_block_ids);
        };

        // `wanted_tenures` was a full reward cycle, so be sure to fetch the tenure-end block of
        // the last tenure derived from it
        if let Some(last_tenure_ch) = last_tenure_ch.take() {
            if let Some(last_tenure) = tenure_block_ids.get_mut(&last_tenure_ch) {
                test_debug!(
                    "Will directly fetch end-block {} for tenure {}",
                    &last_tenure.end_block_id,
                    &last_tenure.tenure_id_consensus_hash
                );
                last_tenure.fetch_end_block = true;
            }
        }

        let Some(next_invbits) = invs.tenures_inv.get(&rc.saturating_add(1)) else {
            // nothing more to do
            test_debug!("no inventory for cycle {}", rc.saturating_add(1));
            return Some(tenure_block_ids);
        };

        // start iterating from the last_tenure
        let iterator = wanted_tenures.get(last_tenure..).unwrap_or(&[]);
        for (i, wt) in iterator.iter().enumerate() {
            test_debug!(
                "consider next wanted tenure which starts with i={} {:?}",
                i,
                &wt
            );

            // advance to next tenure-start sortition
            let bit = u16::try_from(i).expect("FATAL: more sortitions than u16::MAX");
            if !invbits.get(bit).unwrap_or(false) {
                continue;
            }

            let Some((used_next, wt_start_ix, wt_start)) = (i..wanted_tenures.len())
                .find_map(|j| {
                    let bit_index = u16::try_from(j).expect("FATAL: more sortitions than u16::MAX");
                    if invbits.get(bit_index).unwrap_or(false) {
                        wanted_tenures.get(j).map(|tenure| (false, j, tenure))
                    } else {
                        None
                    }
                })
                .or_else(|| {
                    (0..next_wanted_tenures.len()).find_map(|n| {
                        let bit_index =
                            u16::try_from(n).expect("FATAL: more sortitions than u16::MAX");
                        if next_invbits.get(bit_index).unwrap_or(false) {
                            next_wanted_tenures.get(n).map(|tenure| (true, n, tenure))
                        } else {
                            None
                        }
                    })
                })
            else {
                test_debug!("Out of wanted_tenures and next_wanted_tenures");
                break;
            };

            test_debug!("next start tenure is {:?}", &wt_start);

            // find the next 1-bit after that -- corresponds to the tenure-end block ID.
            let next_start = if used_next { wt_start_ix + 1 } else { 0 };
            let Some(wt_end) = (next_start..next_wanted_tenures.len()).find_map(|k| {
                let bit = u16::try_from(k).expect("FATAL: more sortitions than u16::MAX");
                if next_invbits.get(bit).unwrap_or(false) {
                    next_wanted_tenures.get(k)
                } else {
                    None
                }
            }) else {
                test_debug!("Out of next_wanted_tenures");
                break;
            };

            let mut tenure_start_end = TenureStartEnd::new(
                wt.tenure_id_consensus_hash.clone(),
                wt_start.winning_block_id.clone(),
                wt_end.winning_block_id.clone(),
                rc,
                pox_constants
                    .block_height_to_reward_cycle(first_burn_height, wt_start.burn_height)
                    .expect("FATAL: tenure from before system start"),
                wt.processed,
            );
            tenure_start_end.fetch_end_block = true;
            tenure_block_ids.insert(wt.tenure_id_consensus_hash.clone(), tenure_start_end);
        }
        Some(tenure_block_ids)
    }
}
```

This is my rough starting suggestion -- replace the inner loops with finds so that you don't need to have the weird relationship with the mutable indexing variables, use enumerate instead of an index for the outer loops.

Reply:
Sure, will update